[Buttons] Match backgroundColor to titleColor API

romoore opened the pull request:
MDCButton's `backgroundColor:forState:` API will now handle the special
case of `(UIControlStateHighlighted | UIControlStateDisabled)` the same
way that UIButton's `titleColor:forState:` API does. With the exception
of fallback behavior in state handling, their behavior should be
identical.

Context and Background
----------------------

When `UIControlStateDisabled` and `UIControlStateHighlighted` are
included in the same argument for the `state` parameter in
`setTitleColor:forState:`, the `UIControlStateDisabled` bit is ignored.
This means that when setting a `titleColor` for the `(.selected |
.highlighted | .disabled)` state, UIButton will actually store it as the
`(.highlighted | .selected)` state. There is an additional exception
that if there is no current value for the "adjusted" UIControlState,
then no assignment is made.

**Example 1: Fallback to `.highlighted` state**

The following example verifies that UIButton will ignore the
`UIControlStateDisabled` bit in the UIControlState parameter when
combined with `UIControlStateHighlighted`. The result is a "fallback" to
the value for `UIControlStateHighlighted`.

```objc
UIButton *buttonSetHighlighted = [[UIButton alloc] init];
[buttonSetHighlighted setTitleColor:UIColor.orangeColor forState:UIControlStateHighlighted];
XCTAssertEqualObjects([buttonSetHighlighted titleColorForState:(UIControlStateDisabled | UIControlStateHighlighted)],
                      [buttonSetHighlighted titleColorForState:UIControlStateHighlighted]);
XCTAssertEqualObjects([buttonSetHighlighted titleColorForState:(UIControlStateSelected | UIControlStateDisabled | UIControlStateHighlighted)],
                      [buttonSetHighlighted titleColorForState:UIControlStateNormal]);
```

**Example 2: Failure to store value for `(.highlighted | .disabled)`**

The following example verifies that UIButton will discard the assigned
value for `(UIControlStateDisabled | UIControlStateHighlighted)` when no
previous value for `UIControlStateHighlighted` has been set.

```objc
UIButton *buttonSetHighlightedDisabled = [[UIButton alloc] init];
[buttonSetHighlightedDisabled setTitleColor:UIColor.orangeColor forState:(UIControlStateDisabled | UIControlStateHighlighted)];
XCTAssertEqualObjects([buttonSetHighlightedDisabled titleColorForState:(UIControlStateDisabled | UIControlStateHighlighted)],
                      [buttonSetHighlightedDisabled titleColorForState:UIControlStateNormal]);
XCTAssertEqualObjects([buttonSetHighlightedDisabled titleColorForState:(UIControlStateSelected | UIControlStateDisabled | UIControlStateHighlighted)],
                      [buttonSetHighlightedDisabled titleColorForState:UIControlStateNormal]);
```

**Example 3: Assigning value for `(.disabled | .highlighted)` overwrites
previous value for `.highlighted`**

The following example verifies that UIButton will overwrite any previous
value for the `UIControlStateHighlighted` state when a value for
`(UIControlStateDisabled | UIControlStateHighlighted)` is later
assigned.

```objc
UIButton *buttonSetHighlightedDisabled = [[UIButton alloc] init];
[buttonSetHighlightedDisabled setTitleColor:UIColor.purpleColor forState:UIControlStateHighlighted];
[buttonSetHighlightedDisabled setTitleColor:UIColor.orangeColor forState:(UIControlStateDisabled | UIControlStateHighlighted)];
XCTAssertEqualObjects([buttonSetHighlightedDisabled titleColorForState:(UIControlStateDisabled | UIControlStateHighlighted)],
                      UIColor.orangeColor);
XCTAssertEqualObjects([buttonSetHighlightedDisabled titleColorForState:UIControlStateHighlighted],
                      UIColor.orangeColor);
```

Part of #3411

Commit:
[Buttons] Match backgroundColor to titleColor API

MDCButton's `backgroundColor:forState:` API will now handle the special
case of `(UIControlStateHighlighted | UIControlStateDisabled)` the same
way that UIButton's `titleColor:forState:` API does. With the exception
of fallback behavior in state handling, their behavior should be
identical.

Context and Background
----------------------

When `UIControlStateDisabled` and `UIControlStateHighlighted` are
included in the same argument for the `state` parameter in
`setTitleColor:forState:`, the `UIControlStateDisabled` bit is ignored.
This means that when setting a `titleColor` for the `(.selected |
.highlighted | .disabled)` state, UIButton will actually store it as the
`(.highlighted | .selected)` state. There is an additional exception
that if there is no current value for the "adjusted" UIControlState,
then no assignment is made.

**Example 1: Fallback to `.highlighted` state**

The following example verifies that UIButton will ignore the
`UIControlStateDisabled` bit in the UIControlState parameter when
combined with `UIControlStateHighlighted`. The result is a "fallback" to
the value for `UIControlStateHighlighted`.

```objc
 UIButton *buttonSetHighlighted = [[UIButton alloc] init];
  [buttonSetHighlighted setTitleColor:UIColor.orangeColor forState:UIControlStateHighlighted];
  XCTAssertEqualObjects([buttonSetHighlighted titleColorForState:(UIControlStateDisabled | UIControlStateHighlighted)],
                        [buttonSetHighlighted titleColorForState:UIControlStateHighlighted]);
  XCTAssertEqualObjects([buttonSetHighlighted titleColorForState:(UIControlStateSelected | UIControlStateDisabled | UIControlStateHighlighted)],
                        [buttonSetHighlighted titleColorForState:UIControlStateNormal]);
```

**Example 2: Failure to store value for `(.highlighted | .disabled)`**

The following example verifies that UIButton will discard the assigned
value for `(UIControlStateDisabled | UIControlStateHighlighted)` when no
previous value for `UIControlStateHighlighted` has been set.

```objc
UIButton *buttonSetHighlightedDisabled = [[UIButton alloc] init];
  [buttonSetHighlightedDisabled setTitleColor:UIColor.orangeColor forState:(UIControlStateDisabled | UIControlStateHighlighted)];
  XCTAssertEqualObjects([buttonSetHighlightedDisabled titleColorForState:(UIControlStateDisabled | UIControlStateHighlighted)],
                        [buttonSetHighlightedDisabled titleColorForState:UIControlStateNormal]);
  XCTAssertEqualObjects([buttonSetHighlightedDisabled titleColorForState:(UIControlStateSelected | UIControlStateDisabled | UIControlStateHighlighted)],
                        [buttonSetHighlightedDisabled titleColorForState:UIControlStateNormal]);
```

**Example 3: Assigning value for `(.disabled | .highlighted)` overwrites
previous value for `.highlighted`**

The following example verifies that UIButton will overwrite any previous
value for the `UIControlStateHighlighted` state when a value for
`(UIControlStateDisabled | UIControlStateHighlighted)` is later
assigned.

```objc
UIButton *buttonSetHighlightedDisabled = [[UIButton alloc] init];
[buttonSetHighlightedDisabled setTitleColor:UIColor.purpleColor forState:UIControlStateHighlighted];
[buttonSetHighlightedDisabled setTitleColor:UIColor.orangeColor forState:(UIControlStateDisabled | UIControlStateHighlighted)];
XCTAssertEqualObjects([buttonSetHighlightedDisabled titleColorForState:(UIControlStateDisabled | UIControlStateHighlighted)],
                      UIColor.orangeColor);
XCTAssertEqualObjects([buttonSetHighlightedDisabled titleColorForState:UIControlStateHighlighted],
                      UIColor.orangeColor);
```

Part of #3411

material-automation commented on components/Buttons/tests/unit/ButtonsTests.m:
```diff
@@ -345,6 +345,55 @@ - (void)testBackgroundColorForStateUpdatesBackgroundColorWithFallback {
+  for (UIControlState state = 0; state < 8; ++state) {
+    XCTAssertEqualObjects(
+        [testButton backgroundColorForState:state], [uiButton titleColorForState:state],
+        @" for state (%lu)", (unsigned long)state);
```
clang-format suggested the following change:


```
-    XCTAssertEqualObjects(
-        [testButton backgroundColorForState:state], [uiButton titleColorForState:state],
-        @" for state (%lu)", (unsigned long)state);
+    XCTAssertEqualObjects([testButton backgroundColorForState:state],
+                          [uiButton titleColorForState:state], @" for state (%lu)",
+                          (unsigned long)state);
```

<!-- identifier: clang-format -->
<!-- File: components/Buttons/tests/unit/ButtonsTests.m -->

romoore replied:
Done

material-automation commented on components/Buttons/tests/unit/ButtonsTests.m:
```diff
@@ -345,6 +345,55 @@ - (void)testBackgroundColorForStateUpdatesBackgroundColorWithFallback {
+  for (UIControlState state = 0; state < 8; ++state) {
+    XCTAssertEqualObjects(
+                          [testButton backgroundColorForState:state], [uiButton titleColorForState:state],
+                          @" for state (%lu)", (unsigned long)state);
```
clang-format suggested the following change:


```
-    XCTAssertEqualObjects(
-                          [testButton backgroundColorForState:state], [uiButton titleColorForState:state],
-                          @" for state (%lu)", (unsigned long)state);
+    XCTAssertEqualObjects([testButton backgroundColorForState:state],
+                          [uiButton titleColorForState:state], @" for state (%lu)",
+                          (unsigned long)state);
```

<!-- identifier: clang-format -->
<!-- File: components/Buttons/tests/unit/ButtonsTests.m -->

romoore replied:
Done

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

romoore commented:
Testing internally at cl/224152545

Commit:
Clang format

romoore commented:
Internal tests all passed.

randallli reviewed (approved):
Cool looks like you digged deeper than I did. I knew there was some weird state stuff with disabled|highlighted but didn't realize it lost values too.

randallli commented:
just curious: Does this happen for all "forState:" apis? like: 
titleForState:
and imageForState:

romoore commented:
Thanks, @randallli . I would like to get @jverkoey's feedback on this when he has time as well. Since it's an unintuitive bit of behavior, I want to make sure there's alignment about moving toward this behavior across all properties.

romoore commented:
@randallli I double-checked and I believe they do behave the same way.  I have a branch in my local clone that has tests to confirm the behavior.

https://github.com/romoore/material-components-ios/blob/how-buttons-color-state/components/Buttons/tests/unit/ButtonTitleColorAccessibilityMutatorTests.m#L197

Here's a local copy for posterity:

```objc
// Copyright 2018-present the Material Components for iOS authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#import <XCTest/XCTest.h>

#import <UIKit/UIKit.h>

static const UIControlState kDisabledHighlighted =
    (UIControlStateDisabled | UIControlStateHighlighted);

static inline UIControlState StorageStateForState(UIControlState state) {
  if ((state & kDisabledHighlighted) == kDisabledHighlighted) {
    return (state & ~UIControlStateDisabled);
  }
  return state;
}

@interface MYTestButton : NSObject

- (void)setTitleColor:(UIColor *)color forState:(UIControlState)state;
- (UIColor *)titleColorForState:(UIControlState)state;
- (void)setTitle:(NSString *)title forState:(UIControlState)state;
- (NSString *)titleForState:(UIControlState)state;
- (void)setImage:(UIImage *)image forState:(UIControlState)state;
- (UIImage *)imageForState:(UIControlState)state;

@end

@implementation MYTestButton {
  NSMutableDictionary<NSNumber *, UIColor *> *_colors;
  NSMutableDictionary<NSNumber *, NSString *> *_titles;
  NSMutableDictionary<NSNumber *, UIImage *> *_images;
}

- (instancetype)init {
  self = [super init];
  if (self) {
    _colors = [NSMutableDictionary dictionary];
    _colors[@(UIControlStateNormal)] = UIColor.whiteColor;
    _titles = [NSMutableDictionary dictionary];
    _images = [NSMutableDictionary dictionary];
  }
  return self;
}

- (void)setTitleColor:(UIColor *)color forState:(UIControlState)state {
  UIControlState storageState = StorageStateForState(state);

  if (storageState == state || _colors[@(storageState)] != nil) {
    _colors[@(storageState)] = color;
  }
}

- (UIColor *)titleColorForState:(UIControlState)state {
  state = StorageStateForState(state);
  return _colors[@(state)] ?: _colors[@(UIControlStateNormal)];
}

- (void)setTitle:(NSString *)title forState:(UIControlState)state {
  UIControlState storageState = StorageStateForState(state);

  if (storageState == state || _titles[@(storageState)] != nil) {
    _titles[@(storageState)] = title;
  }
}

- (NSString *)titleForState:(UIControlState)state {
  state = StorageStateForState(state);
  return _titles[@(state)] ?: _titles[@(UIControlStateNormal)];
}

- (void)setImage:(UIImage *)image forState:(UIControlState)state {
  UIControlState storageState = StorageStateForState(state);

  if (storageState == state || _images[@(storageState)] != nil) {
    _images[@(storageState)] = image;
  }
}

- (UIImage *)imageForState:(UIControlState)state {
  state = StorageStateForState(state);
  return _images[@(state)] ?: _images[@(UIControlStateNormal)];
}

@end

@interface UIButtonBehaviorTests : XCTestCase

@end

@implementation UIButtonBehaviorTests

- (void)testTitleColorForStateImplementation {

  for (NSInteger lowestAssignState = 7; lowestAssignState >= 0; --lowestAssignState) {
    MYTestButton *testButton = [[MYTestButton alloc] init];
    UIButton *uiButton = [[UIButton alloc] init];

    for (UIControlState assignState = (NSUInteger)lowestAssignState; assignState <= 7; ++assignState) {
      [testButton setTitleColor:[UIColor colorWithWhite:(CGFloat)(lowestAssignState / 8.0) alpha:(CGFloat)(assignState / 8.0)]
                       forState:assignState];
      [uiButton setTitleColor:[UIColor colorWithWhite:(CGFloat)(lowestAssignState / 8.0) alpha:(CGFloat)(assignState / 8.0)]
                     forState:assignState];
    }
    for (UIControlState testState = 0; testState <= 7; ++testState ) {
      XCTAssertEqualObjects([testButton titleColorForState:testState],
                            [uiButton titleColorForState:testState],
                            @" for Lowest = (%ld), Test = (%lu)",
                            lowestAssignState, testState);
    }
  }
}

- (void)testTitleForStateImplementation {

  for (NSInteger lowestAssignState = 7; lowestAssignState >= 0; --lowestAssignState) {
    MYTestButton *testButton = [[MYTestButton alloc] init];
    UIButton *uiButton = [[UIButton alloc] init];

    for (UIControlState assignState = (NSUInteger)lowestAssignState; assignState <= 7; ++assignState) {
      NSString *string = [NSString stringWithFormat:@"%lu %lu", (unsigned long)lowestAssignState, (unsigned long)assignState];
      [testButton setTitle:string forState:assignState];
      [uiButton setTitle:string forState:assignState];
    }
    for (UIControlState testState = 0; testState <= 7; ++testState ) {
      XCTAssertEqualObjects([testButton titleForState:testState],
                            [uiButton titleForState:testState],
                            @" for Lowest = (%ld), Test = (%lu)",
                            lowestAssignState, testState);
    }
  }
}

- (void)testImageForStateImplementation {

  for (NSInteger lowestAssignState = 7; lowestAssignState >= 0; --lowestAssignState) {
    MYTestButton *testButton = [[MYTestButton alloc] init];
    UIButton *uiButton = [[UIButton alloc] init];

    for (UIControlState assignState = (NSUInteger)lowestAssignState; assignState <= 7; ++assignState) {
      UIImage *image = [[UIImage alloc] init];
      [testButton setImage:image forState:assignState];
      [uiButton setImage:image forState:assignState];
    }
    for (UIControlState testState = 0; testState <= 7; ++testState ) {
      XCTAssertEqual([testButton imageForState:testState],
                     [uiButton imageForState:testState],
                     @" for Lowest = (%ld), Test = (%lu)",
                     lowestAssignState, testState);
    }
  }
}

@end

```

fluxminds commented on components/Buttons/tests/unit/ButtonsTests.m:
```diff
@@ -345,6 +345,55 @@ - (void)testBackgroundColorForStateUpdatesBackgroundColorWithFallback {
+  UIButton *uiButton = [[UIButton alloc] init];
+
+  // When
+  for (UIControlState state = 0; state < 8; ++state) {
```
Trivial: using `UIControlState maxState = UIControlStateNormal | UIControlStateHighlighted | UIControlStateDisabled | UIControlStateSelected;` would help with readability.

romoore replied:
Done.

What does the "Trivial:" heading mean? That's a new one for me.

fluxminds replied:
I meant _Nit_.

romoore replied:
Makes sense, thanks!

Commit:
Merge branch 'develop' into match-uicolor-states

Commit:
Tests: Clarify UIControlState-based loop

The loops used for testing would iterate in [0, 8). Since the
combination of all UIKit-based states (.selected, .disabled,
.highlighted) is 7, the loops were going one more than that. By
bitwise-OR'ing the states together the tests can be a bit clearer about
what they're doing. The new loop range is [0, 7].

material-automation commented on components/Buttons/tests/unit/ButtonsTests.m:
```diff
@@ -345,6 +345,59 @@ - (void)testBackgroundColorForStateUpdatesBackgroundColorWithFallback {
+
+  // When
+  UIControlState maxState = UIControlStateNormal | UIControlStateHighlighted
+      | UIControlStateDisabled | UIControlStateSelected;
```
clang-format suggested the following change:


```
-  UIControlState maxState = UIControlStateNormal | UIControlStateHighlighted
-      | UIControlStateDisabled | UIControlStateSelected;
+  UIControlState maxState = UIControlStateNormal | UIControlStateHighlighted |
+                            UIControlStateDisabled | UIControlStateSelected;
```

<!-- identifier: clang-format -->
<!-- File: components/Buttons/tests/unit/ButtonsTests.m -->

romoore replied:
Done.

material-automation commented on components/Buttons/tests/unit/ButtonsTests.m:
```diff
@@ -345,6 +345,59 @@ - (void)testBackgroundColorForStateUpdatesBackgroundColorWithFallback {
+
+  // When
+  UIControlState maxState = UIControlStateNormal | UIControlStateHighlighted
+      | UIControlStateDisabled | UIControlStateSelected;
```
clang-format suggested the following change:


```
-  UIControlState maxState = UIControlStateNormal | UIControlStateHighlighted
-      | UIControlStateDisabled | UIControlStateSelected;
+  UIControlState maxState = UIControlStateNormal | UIControlStateHighlighted |
+                            UIControlStateDisabled | UIControlStateSelected;
```

<!-- identifier: clang-format -->
<!-- File: components/Buttons/tests/unit/ButtonsTests.m -->

romoore replied:
Done.

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

Commit:
Clang formatting for comment feedback

Commit:
Merge branch 'develop' into match-uicolor-states

Commit:
Simplify and document logic.

Commit:
Fix color-assignment logic broken by last refactor.

jverkoey reviewed (approved):
Solid sleuthing!
